src/render/ngage/SDL_render_ngage.cpp

```diff
@@ -0,0 +1,790 @@
+    }
+}
+
+void CRenderer::SetClipRect(TInt aX, TInt aY, TInt aWidth, TInt aHeight)
+{
+    if (iRenderer && iRenderer->Gc())
+    {
+        TRect viewportRect(aX, aY, aX + aWidth, aY + aHeight);
+        iRenderer->Gc()->SetClippingRect(viewportRect);
+    }
+}
+
+void CRenderer::UpdateFPS()
+{
+    static TTime lastTime;
+    static TInt frameCount = 0;
+    TTime currentTime;
+    const TUint KOneSecond = 1000000; // 1s in ms.
+
+    currentTime.HomeTime();
+    ++frameCount;
+
+    TTimeIntervalMicroSeconds timeDiff = currentTime.MicroSecondsFrom(lastTime);
+
+    if (timeDiff.Int64() >= KOneSecond)
+    {
+        // Calculate FPS.
+        iFPS = frameCount;
+
+        // Reset frame count and last time.
+        frameCount = 0;
+        lastTime = currentTime;
+    }
+}
+
+static SDL_Scancode ConvertScancode(int key)
+{
+    SDL_Keycode keycode;
+
+    switch(key)
+    {
+        case EStdKeyBackspace: // Clear key
+            keycode = SDLK_BACKSPACE;
+            break;
+        case 0x31: // 1
+            keycode = SDLK_1;
+            break;
+        case 0x32: // 2
+            keycode = SDLK_2;
+            break;
+        case 0x33: // 3
+            keycode = SDLK_3;
+            break;
+        case 0x34: // 4
+            keycode = SDLK_4;
+            break;
+        case 0x35: // 5
+            keycode = SDLK_5;
+            break;
+        case 0x36: // 6
+            keycode = SDLK_6;
+            break;
+        case 0x37: // 7
+            keycode = SDLK_7;
+            break;
+        case 0x38: // 8
+            keycode = SDLK_8;
+            break;
+        case 0x39: // 9
+            keycode = SDLK_9;
+            break;
+        case 0x30: // 0
+            keycode = SDLK_0;
+            break;
+        case 0x2a: // Asterisk
+            keycode = SDLK_ASTERISK;
+            break;
+        case EStdKeyHash: // Hash
+            keycode = SDLK_HASH;
+            break;
+        case EStdKeyDevice0: // Left softkey
+            keycode = SDLK_SOFTLEFT;
+            break;
+        case EStdKeyDevice1: // Right softkey
+            keycode = SDLK_SOFTRIGHT;
+            break;
+        case EStdKeyApplication0: // Call softkey
+            keycode = SDLK_CALL;
+            break;
+        case EStdKeyApplication1: // End call softkey
+            keycode = SDLK_ENDCALL;
+            break;
+        case EStdKeyDevice3: // Middle softkey
+            keycode = SDLK_SELECT;
+            break;
+        case EStdKeyUpArrow: // Up arrow
+            keycode = SDLK_UP;
+            break;
+        case EStdKeyDownArrow: // Down arrow
+            keycode = SDLK_DOWN;
+            break;
+        case EStdKeyLeftArrow: // Left arrow
+            keycode = SDLK_LEFT;
+            break;
+        case EStdKeyRightArrow: // Right arrow
+            keycode = SDLK_RIGHT;
+            break;
+        default:
+            keycode = SDLK_UNKNOWN;
+            break;
+    }
+
+    return SDL_GetScancodeFromKey(keycode, NULL);
```

Comment:
I'd suggest directly using the `SDL_SCANCODE_` values here rather than converting from a fixed set of `SDLK_` values.

Reply:
How do you mean that? I am currently translating the scan codes from the Symbian API into the identifiers used by SDL.

Reply:
I would rather set a variable while having a single point of return, however, I need to determine all the scan codes to make this work either way. 👍